qopt_best_practices/transpilation/qaoa_construction_pass.py

```diff
@@ -0,0 +1,124 @@
+from qiskit.transpiler import TranspilerError
+from qiskit.transpiler.basepasses import TransformationPass
+
+
+class QAOAConstructionPass(TransformationPass):
+    """Build the QAOAAnsatz from a transpiled cost operator.
+
+    This pass takes as input a single layer of a transpiled QAOA operator.
+    It then repeats this layer the appropriate number of time and adds (i)
+    the initial state, (ii) the mixer operator, and (iii) the measurements.
+    The measurements are added in such a fashion so as to undo the local
+    permuttion induced by the SWAP gates in the cost layer.
+    """
+
+    def __init__(
+        self,
+        num_layers: int,
+        init_state: Optional[QuantumCircuit] = None,
+        mixer_layer: Optional[QuantumCircuit] = None,
+    ):
+        """Initialize the pass
+
+        Limitations: The current implementation of the pass does not permute the mixer.
+        Therefore mixers with local bias fields, such as in warm-start methods, will not
+        result in the correct circuits.
+
+        Args:
+            num_layers: The number of QAOA layers to apply.
+            init_state: The initial state to use. This must match the anticipated number
+                of qubits otherwise an error will be raised when `run` is called. If this
+                is not given we will default to the equal superposition initial state.
+            mixer_layer: The mixer layer to use. This must match the anticipated number
+                of qubits otherwise an error will be raised when `run` is called. If this
+                is not given we default to the standard mixer made of X gates.
+        """
+        super().__init__()
+
+        self.num_layers = num_layers
+        self.init_state = init_state
+        self.mixer_layer = mixer_layer
+
+    def run(self, cost_layer_dag):
+        num_qubits = cost_layer_dag.num_qubits()
+
+        # Make the initial state and the mixer.
+        if self.init_state is None:
+            init_state = QuantumCircuit(num_qubits)
+            init_state.h(range(num_qubits))
+        else:
+            init_state = self.init_state
+
+        if self.mixer_layer is None:
+            mixer_layer = QuantumCircuit(num_qubits)
+            beta = Parameter("β")
+            mixer_layer.rx(2 * beta, range(num_qubits))
+        else:
+            mixer_layer = self.mixer_layer
+
+        # Do some sanity checks on qubit numbers.
+        if init_state.num_qubits != num_qubits:
+            raise TranspilerError(
+                "Number of qubits in the initial state does not match the number in the DAG. "
+                f"{init_state.num_qubits} != {num_qubits}"
+            )
+
+        if mixer_layer.num_qubits != num_qubits:
+            raise TranspilerError(
+                "Number of qubits in the initial state does not match the number in the DAG. "
+                f"{init_state.num_qubits} != {num_qubits}"
+            )
+
+        cost_layer = dag_to_circuit(cost_layer_dag)
```

Comment:
I am fine keeping this as a first implementation but noting that in the future it would be more performant not to convert to circuit here and instead, work with a dag end-to-end.

Reply:
Agreed. I left a comment to this effect. See here: [0ae273b](https://github.com/qiskit-community/qopt-best-practices/pull/34/commits/0ae273bec6cbe52f27ad9d27006e52227ecf45e7)